### PR TITLE
TagBox: fix tags selection when items are objects, valueExpr="this" and applyValueMode='useButtons' (T1141799)

### PR DESCRIPTION
**Pull request by AlexanderMoiseev**

### CODE REVIEW

**Comment by ksercs** [comment]
@AlexanderMoiseev  
1) remove excess spaces in ID brackets plz. It'll simplify future searches
2) name is too common for me. I read it and it seems like tags selection does not work at all.. But it works, we fix only very specific case. Can we describe it in details?

File: js/core/utils/array.js [hunk 1]
```diff
@@ -3,10 +3,16 @@ import { orderEach } from './object';
 import config from '../config';
 
 function createOccurrenceMap(array) {
-    return array.reduce((map, value) => {
-        map[value] = (map[value] ?? 0) + 1;
-        return map;
-    }, {});
+    return array.reduce((map, value) => map.set(value, (map.get(value) ?? 0) + 1), new Map());
```
Comment (ksercs) [inline js/core/utils/array.js]: It is really difficult to read. I think that readability is more important than shortness
```suggestion
    return array.reduce((map, value) => {
        const count = (map.get(value) ?? 0) + 1;
        map.set(value, count);
        return map;
    }, new Map());
```

File: js/core/utils/array.js [hunk 2]
```diff
@@ -3,10 +3,16 @@ import { orderEach } from './object';
...
+}
+
+function getFilteredArray(from, toRemove, getUniqueItems) {
+    const toRemoveMap = createOccurrenceMap(toRemove);
+    return from.filter(value => {
+        const occurrencesCount = toRemoveMap.get(value);
+        occurrencesCount && toRemoveMap.set(value, occurrencesCount - 1);
+        return getUniqueItems ^ occurrencesCount;
```
Comment (ksercs) [inline js/core/utils/array.js]: To me this function only worsens a readability. Why did u decide to separate it?
- it increases count of code rows
- it needs to pass a non-intuitive true/false parameter
- it uses binary operator
- we don't see simple filtering in parent functions

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 3]
```diff
@@ -5543,6 +5543,52 @@ QUnit.module('applyValueMode = \'useButtons\'', {
         assert.deepEqual(this.instance.option('value'), [items[2], items[0]], 'tags order is correct');
     });
 
+    QUnit.test('Object value should be updated correctly when item is added', function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]:
```suggestion
    QUnit.test('value should be updated correctly after item is added if valueExpr="this" (T1141799)', function(assert) {
```

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 4]
```diff
@@ -5543,6 +5543,52 @@ QUnit.module('applyValueMode = \'useButtons\'', {
...
+        $(this.$popupWrapper.find(`.${POPUP_DONE_BUTTON_CLASS}`)).trigger('dxclick');
+
+        const items = this.instance.option('value');
+        assert.equal(items.length, 3);
+        assert.deepEqual(items[0], firstValue);
+        assert.deepEqual(items[1], thirdValue);
+        assert.deepEqual(items[2], secondValue);
+    });
+
+    QUnit.test('Object value should be updated correctly when item is removed', function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]:
```suggestion
    QUnit.test('value should be updated correctly after item is removed if valueExpr="this" (T1141799)', function(assert) {
```

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 5]
```diff
@@ -5543,6 +5543,52 @@ QUnit.module('applyValueMode = \'useButtons\'', {
...
+            applyValueMode: 'useButtons',
+            displayExpr: 'description',
+            opened: true
+        });
+
+        $(this.$listItems.eq(1)).trigger('dxclick');
+        $(this.$popupWrapper.find(`.${POPUP_DONE_BUTTON_CLASS}`)).trigger('dxclick');
+
+        const items = this.instance.option('value');
+        assert.equal(items.length, 3);
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]:
```suggestion
        assert.strictEqual(items.length, 3);
```

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 6]
```diff
@@ -5543,6 +5543,52 @@ QUnit.module('applyValueMode = \'useButtons\'', {
...
+            applyValueMode: 'useButtons',
+            displayExpr: 'description',
+            opened: true
+        });
+
+        $(this.$listItems.eq(1)).trigger('dxclick');
+        $(this.$popupWrapper.find(`.${POPUP_DONE_BUTTON_CLASS}`)).trigger('dxclick');
+
+        const items = this.instance.option('value');
+        assert.equal(items.length, 2);
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]:
```suggestion
        assert.strictEqual(items.length, 2);
```

File: js/core/utils/array.js (hunk 2 above)
Comment (AlexanderMoiseev) [reply js/core/utils/array.js]: 1) "Why did u decide to separate it?" - i extracted this function to reuse logic. ( look at the 2-nd note below )
2) "it increases count of code rows" -  Not sure i follow this  argument. Actually it does the opposite ) or did you mean something else? but in regard to the parameter and operator- yes that way its more readable. i changed it, please take a look

Comment (ksercs) [reply js/core/utils/array.js]: oh, Alex, i didn't note that we use Map instead of object => code copypaste is much wider that was before, sorry

Comment (ksercs) [reply js/core/utils/array.js]: But i think actual variant is ok anyway

**Review by ksercs — APPROVED** [review]

**Review by marker-dao — APPROVED** [review]

**Pull request merged**